// @streamdal/node-sdk/cjs/sandbox/billing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onboardingExample = exports.singleWelcomeExample = exports.welcomeExample = exports.singleSignupExample = exports.signupExample = exports.billingExample = exports.randomPipelineAndData = exports.loadData = void 0;
const node_fs_1 = require("node:fs");
const index_js_1 = require("../index.js");
const index_js_2 = require("./index.js");
const serviceBillingConfig = {
    streamdalUrl: "localhost:8082",
    streamdalToken: "1234",
    serviceName: "billing-service",
    pipelineTimeout: "100",
    stepTimeout: "10",
    dryRun: false,
};
const serviceSignupConfig = {
    streamdalUrl: "localhost:8082",
    streamdalToken: "1234",
    serviceName: "signup-service",
    pipelineTimeout: "100",
    stepTimeout: "10",
    dryRun: false,
};
const serviceWelcomeConfig = {
    streamdalUrl: "localhost:8082",
    streamdalToken: "1234",
    serviceName: "welcome-service",
    pipelineTimeout: "100",
    stepTimeout: "10",
    dryRun: false,
};
const billingConsumer = {
    serviceName: "billing-service",
    componentName: "kafka",
    operationType: index_js_1.OperationType.CONSUMER,
    operationName: "processor",
};
const billingProducer = {
    serviceName: "billing-service",
    componentName: "stripe",
    operationType: index_js_1.OperationType.PRODUCER,
    operationName: "stripe-register",
};
const signupProducer = {
    serviceName: "signup-service",
    componentName: "kafka",
    operationType: index_js_1.OperationType.PRODUCER,
    operationName: "recorder",
};
const signupProducer1 = {
    serviceName: "signup-service",
    componentName: "database",
    operationType: index_js_1.OperationType.PRODUCER,
    operationName: "verifier",
};
const welcomeConsumer = {
    serviceName: "welcome-service",
    componentName: "kafka",
    operationType: index_js_1.OperationType.CONSUMER,
    operationName: "reader",
};
const welcomeProducer = {
    serviceName: "welcome-service",
    componentName: "aws-ses",
    operationType: index_js_1.OperationType.PRODUCER,
    operationName: "emailer",
};
const parseJson = (d) => {
    try {
        return d ? JSON.parse(d) : null;
    }
    catch (e) {
        console.error("error parsing sample data to json", e);
    }
    return null;
};
const loadData = (path) => {
    try {
        const data = (0, node_fs_1.readFileSync)(path).toString();
        const parsed = data
            .split(/\r?\n/)
            .map((d) => parseJson(d))
            .filter((d) => d);
        return parsed;
    }
    catch (e) {
        console.error("error loading sample data", e);
    }
    return [];
};
exports.loadData = loadData;
const randomPipelineAndData = (streamdal, audience, input) => {
    (0, index_js_2.runPipeline)(streamdal, audience, input[Math.floor(Math.random() * input.length)]);
    setTimeout(() => (0, exports.randomPipelineAndData)(streamdal, audience, input), Math.floor(Math.random() * (3000 - 100) + 100));
};
exports.randomPipelineAndData = randomPipelineAndData;
const billingExample = async () => {
    const streamdal = await (0, index_js_1.registerStreamdal)({
        ...serviceBillingConfig,
        quiet: index_js_2.QUIET,
    });
    const bcData = (0, exports.loadData)("./src/sandbox/assets/sample-billing-consumer.json");
    const bpData = (0, exports.loadData)("./src/sandbox/assets/sample-billing-producer.json");
    (0, exports.randomPipelineAndData)(streamdal, billingConsumer, bcData);
    (0, exports.randomPipelineAndData)(streamdal, billingProducer, bpData);
};
exports.billingExample = billingExample;
const signupExample = async () => {
    const streamdal = await (0, index_js_1.registerStreamdal)({
        ...serviceSignupConfig,
        quiet: index_js_2.QUIET,
    });
    const sData = (0, exports.loadData)("./src/sandbox/assets/sample.json");
    const spData = (0, exports.loadData)("./src/sandbox/assets/sample-signup-producer.json");
    (0, exports.randomPipelineAndData)(streamdal, signupProducer, spData);
    (0, exports.randomPipelineAndData)(streamdal, signupProducer1, sData);
};
exports.signupExample = signupExample;
const singleSignupExample = async () => {
    const streamdal = await (0, index_js_1.registerStreamdal)({
        ...serviceSignupConfig,
        quiet: index_js_2.QUIET,
    });
    const spData = (0, exports.loadData)("./src/sandbox/assets/sample-signup-producer.json");
    (0, exports.randomPipelineAndData)(streamdal, signupProducer, spData);
};
exports.singleSignupExample = singleSignupExample;
const welcomeExample = async () => {
    const streamdal = await (0, index_js_1.registerStreamdal)(serviceWelcomeConfig);
    const sData = (0, exports.loadData)("./src/sandbox/assets/sample.json");
    const wpData = (0, exports.loadData)("./src/sandbox/assets/sample-welcome-producer.json");
    (0, exports.randomPipelineAndData)(streamdal, welcomeConsumer, sData);
    (0, exports.randomPipelineAndData)(streamdal, welcomeProducer, wpData);
};
exports.welcomeExample = welcomeExample;
const singleWelcomeExample = async () => {
    const streamdal = await (0, index_js_1.registerStreamdal)({
        ...serviceWelcomeConfig,
        quiet: index_js_2.QUIET,
    });
    const wpData = (0, exports.loadData)("./src/sandbox/assets/sample-welcome-producer.json");
    (0, exports.randomPipelineAndData)(streamdal, welcomeProducer, wpData);
};
exports.singleWelcomeExample = singleWelcomeExample;
const onboardingExample = () => {
    void (0, exports.signupExample)();
    void (0, exports.billingExample)();
    void (0, exports.welcomeExample)();
};
exports.onboardingExample = onboardingExample;
