// @streamdal/node-sdk/cjs/internal/utils/tokenBucket.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenBucket = exports.SAMPLE_DEFAULT = void 0;
exports.SAMPLE_DEFAULT = {
    sampleRate: 25,
    sampleIntervalSeconds: 1,
};
class TokenBucket {
    sampleRate;
    tokens;
    sampleIntervalSeconds;
    lastRefill;
    constructor(sampleRate, sampleIntervalSeconds) {
        this.sampleRate = sampleRate ?? exports.SAMPLE_DEFAULT.sampleRate;
        this.sampleIntervalSeconds =
            (sampleIntervalSeconds ?? exports.SAMPLE_DEFAULT.sampleIntervalSeconds) / 1000;
        this.lastRefill = Math.floor(Date.now() / 1000);
        this.tokens = this.sampleRate;
    }
    refill() {
        const now = Math.floor(Date.now() / 1000);
        const rate = (now - this.lastRefill) / this.sampleIntervalSeconds;
        this.tokens = Math.min(this.sampleRate, this.tokens + Math.floor(rate * this.sampleRate));
        this.lastRefill = now;
    }
    consume() {
        this.refill();
        if (this.tokens > 0) {
            this.tokens -= 1;
            return true;
        }
        return false;
    }
}
exports.TokenBucket = TokenBucket;
