// @streamdal/node-sdk/cjs/internal/httpRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpRequest = void 0;
const sp_wsm_1 = require("@streamdal/protos/protos/sp_wsm");
const sp_steps_httprequest_1 = require("@streamdal/protos/protos/steps/sp_steps_httprequest");
const httpRequest = async ({ step }) => {
    if (step.step.oneofKind !== "httpRequest" || !step.step.httpRequest.request) {
        throw new Error("not a valid httpRequest");
    }
    const response = await fetch(step.step.httpRequest.request.url, {
        method: sp_steps_httprequest_1.HttpRequestMethod[step.step.httpRequest.request.method],
        body: new TextDecoder().decode(step.step.httpRequest.request.body),
        headers: step.step.httpRequest.request.headers,
    });
    let body;
    try {
        body = await response.json();
    }
    catch {
        console.debug("could not parse http request response body");
    }
    return {
        outputStep: null,
        outputPayload: new Uint8Array(),
        exitCode: response.ok
            ? sp_wsm_1.WASMExitCode.WASM_EXIT_CODE_TRUE
            : sp_wsm_1.WASMExitCode.WASM_EXIT_CODE_ERROR,
        ...(body ? { exitMsg: JSON.stringify(body) } : {}),
        interStepResult: undefined,
    };
};
exports.httpRequest = httpRequest;
